src/basic/mkdir.c

```diff
@@ -248,16 +252,28 @@ int mkdir_p_root(const char *root, const char *p, uid_t uid, gid_t gid, mode_t m
                 return r;
         }
 
```

Comment:
Just return early here if `ts == USEC_INFINITY && !uid_is_valid(uid) && !gid_is_valid(gid)`? Then, move declaration of `nfd` below so that it's more obvious when it would be used.